pkg/inclusion/nmt_caching.go

```diff
@@ -0,0 +1,125 @@
+	}
+
+	// continue to traverse the tree by recursively calling this function on the next root
+	switch path[0] {
+	// walk left
+	case false:
+		return strc.walk([]byte(children[0]), path[1:])
+	// walk right
+	case true:
+		return strc.walk([]byte(children[1]), path[1:])
+	default:
+		// this is unreachable code, but the compiler doesn't recognize this somehow
+		panic("bool other than true or false, computers were a mistake, everything is a lie, math is fake.")
+	}
+}
+
+// EDSSubTreeRootCacher caches the inner nodes for each row so that we can
+// traverse it later to check for message inclusion. NOTE: Currently this has to
+// use a leaky abstraction (see docs on counter field below), and is not
+// threadsafe, but with a future refactor, we could simply read from rsmt2d and
+// not use the tree constructor which would fix both of these issues.
+type EDSSubTreeRootCacher struct {
+	caches     []*subTreeRootCacher
+	squareSize uint64
+	// counter is used to ignore columns NOTE: this is a leaky abstraction that
+	// we make because rsmt2d is used to generate the roots for us, so we have
+	// to assume that it will generate a row root every other tree contructed.
+	// This is also one of the reasons this implementation is not thread safe.
+	// Please see note above on a better refactor.
+	counter int
+}
+
+func NewCachedSubtreeCacher(squareSize uint64) *EDSSubTreeRootCacher {
+	return &EDSSubTreeRootCacher{
+		caches:     []*subTreeRootCacher{},
+		squareSize: squareSize,
+	}
+}
+
+// Constructor fullfills the rsmt2d.TreeCreatorFn by keeping a pointer to the
+// cache and embedding it as a nmt.NodeVisitor into a new wrapped nmt. I only
+func (stc *EDSSubTreeRootCacher) Constructor() rsmt2d.Tree {
+	// see docs of counter field for more
+	// info. if the counter is even or == 0, then we make the assumption that we
+	// are creating a tree for a row
+	var newTree wrapper.ErasuredNamespacedMerkleTree
+	switch stc.counter % 2 {
+	case 0:
+		strc := newSubTreeRootCacher()
+		stc.caches = append(stc.caches, strc)
+		newTree = wrapper.NewErasuredNamespacedMerkleTree(stc.squareSize, nmt.NodeVisitor(strc.Visit))
+	default:
+		newTree = wrapper.NewErasuredNamespacedMerkleTree(stc.squareSize)
+	}
+
+	stc.counter++
+	return &newTree
+}
+
+// GetSubTreeRoot traverses the nmt of the selected row and returns the
+// subtree root. An error is thrown if the subtree cannot be found.
+func (stc *EDSSubTreeRootCacher) GetSubTreeRoot(dah da.DataAvailabilityHeader, row int, path []bool) ([]byte, error) {
+	const unexpectedDAHErr = "data availability header has unexpected number of row roots: expected %d got %d"
+	if len(stc.caches) != len(dah.RowsRoots) {
+		return nil, fmt.Errorf(unexpectedDAHErr, len(stc.caches), len(dah.RowsRoots))
+	}
+	const rowOutOfBoundsErr = "row exceeds range of cache: max %d got %d"
+	if row >= len(stc.caches) {
+		return nil, fmt.Errorf(rowOutOfBoundsErr, len(stc.caches), row)
+	}
```

Comment:
[question][non-blocking] given these error messages are only used once, what is the benefit to extracting constants over inlining the message like:

```suggestion
	if len(stc.caches) != len(dah.RowsRoots) {
		return nil, fmt.Errorf("data availability header has unexpected number of row roots: expected %d got %d", len(stc.caches), len(dah.RowsRoots))
	}
	if row >= len(stc.caches) {
		return nil, fmt.Errorf("row exceeds range of cache: max %d got %d", len(stc.caches), row)
	}
```